[core] Execute the pipe-processors in their initialization order

flaviendelangle opened the pull request:
## Problem

While working on the aggregation (#4208) I noticed that the execution order of the pipe-processors was from the oldest-updated-processor to the latest-updated-processor.
We should try to always have execution orders respect the order of initialization (ie the order of the hooks in `useDataGridXXXComponent`).

This unstable execution order was causing a double execution of `hydrateColumns` in some scenario when used with aggregation and row grouping.

### Reproduction case

If you change a prop triggering an update of the row grouping processor of `hydrateColumns` (for instance switch `rowGroupingColumnModel` from `single` to `multiple`), then the aggregation processor of `hydrateColumns` will be ran before the row grouping one.
But the aggregation needs the new row grouping columns to decide how to aggregate.
So we have 2 execution of `hydrateColumns`

1. The `hydrateColumns` processor of Row Grouping is updated, firing a 1st `hydrateColumns` execution
2. `hydrateColumns` processor executed: Aggregation does nothing, Row Grouping update its columns
3. `columnsChange` events is fired, Aggregation manually fires a 2nd `hydrateColumns` execution because it needs to update its columns
4. `hydrateColumns` processor executed: Aggregation updates its columns, Row Grouping does nothing

### Expected behavior

1. The `hydrateColumns` processor of Row Grouping is updated, firing a 1st `hydrateColumns` execution
2. `hydrateColumns` processor executed:  Row Grouping update its columns, Aggregation updates its columns based on the new Row Grouping columns,
3. `columnsChange` events is fired, Aggregation do not fires a 2nd `hydrateColumns` because its columns are up to date

## Solution

- [x] Replace the object storing the processors by a `Map` to have a guaranteed key order
- [x] Do not delete the map item when removing a processor but replace it by `null` to keep its position in the execution order

cherniavskii commented:
> Do not delete the map item when removing a processor but replace it by null to keep its position in the execution order

Is there a use case behind it? If I understand it correctly, it would only matter if processor is added, then removed and added again?

flaviendelangle commented:
> it would only matter if processor is added, then removed and added again?

The processor are added in a `useEffect` and removed in the cleanup function
So they are removed then re-added.

But we don't listen to processor removal to execute them (otherwise we would execute a lot of thing twice)
Not sure how to clean that in the future. But we do need to support removal in the ordering :+1:
